cpp/src/fil/common.cuh

```diff
@@ -173,6 +172,101 @@ struct predict_params : shmem_size_params {
   int num_blocks;
 };
 
+template <bool COLS_IN_SHMEM, int LEAF_ALGO, int N_ITEMS>
+struct KernelTemplateParameters {
+  static const bool cols_in_shmem = COLS_IN_SHMEM;
+  static const leaf_algo_t leaf_algo = (leaf_algo_t)LEAF_ALGO;
+  static const int n_items = N_ITEMS;
+};
+
+namespace dispatch {
+
+template <template <class> class Func, class KernelParams, typename... Args>
+void dispatch_on_n_items(predict_params& params, Args... args) {
+  if (params.n_items == KernelParams::n_items) {
+    Func<KernelParams>(params, args...);
+  } else if constexpr (KernelParams::n_items < 4) {
+    typedef KernelTemplateParameters<KernelParams::cols_in_shmem,
```

Comment:
Repeating all fields each time doesn't look nice. Isn't it possible to define a producer method/template/type that copies all constants in `KernelTemplateParameters` except for one for which the new value is specified?

Alternatively, it could just select the next value of the parameter.

Perhaps for this purpose, the recursive definition of `KernelParams` would work better.